sycl/plugins/unified_runtime/ur/adapters/cuda/image.cpp

```diff
@@ -0,0 +1,1029 @@
+  UR_ASSERT(hImageMem, UR_RESULT_ERROR_INVALID_NULL_HANDLE);
+  UR_ASSERT(pImageFormat, UR_RESULT_ERROR_INVALID_NULL_POINTER);
+  UR_ASSERT(pImageDesc, UR_RESULT_ERROR_INVALID_NULL_POINTER);
+  UR_ASSERT(phMem, UR_RESULT_ERROR_INVALID_NULL_POINTER);
+  UR_ASSERT(phImage, UR_RESULT_ERROR_INVALID_NULL_POINTER);
+
+  ur_result_t RetErr = UR_RESULT_SUCCESS;
+  ScopedContext Active(hDevice->getContext());
+
+  unsigned int NumChannels = 0;
+  RetErr = UR_CHECK_ERROR(
+      urCalculateNumChannels(pImageFormat->channelOrder, &NumChannels));
+
+  CUarray_format format;
+  size_t PixelTypeSizeBytes;
+  RetErr = UR_CHECK_ERROR(urToCudaImageChannelFormat(
+      pImageFormat->channelType, &format, &PixelTypeSizeBytes));
+
+  try {
+    CUDA_RESOURCE_DESC image_res_desc = {};
+
+    unsigned int mem_type;
+    // If this function doesn't return successfully, we assume that hImageMem is
+    // a CUarray or CUmipmappedArray. If this function returns successfully, we
+    // check whether hImageMem is device memory (even managed memory isn't
+    // considered shared).
+    CUresult Err = cuPointerGetAttribute(
+        &mem_type, CU_POINTER_ATTRIBUTE_MEMORY_TYPE, (CUdeviceptr)hImageMem);
+    if (Err != CUDA_SUCCESS) {
+      // We have a CUarray
+      if (pImageDesc->numMipLevel == 1) {
+        image_res_desc.resType = CU_RESOURCE_TYPE_ARRAY;
+        image_res_desc.res.array.hArray = (CUarray)hImageMem;
+      }
+      // We have a CUmipmappedArray
+      else {
+        image_res_desc.resType = CU_RESOURCE_TYPE_MIPMAPPED_ARRAY;
+        image_res_desc.res.mipmap.hMipmappedArray = (CUmipmappedArray)hImageMem;
+      }
+    } else if (mem_type == CU_MEMORYTYPE_DEVICE) {
+      // We have a USM pointer
+      if (pImageDesc->type == UR_MEM_TYPE_IMAGE1D) {
+        image_res_desc.resType = CU_RESOURCE_TYPE_LINEAR;
+        image_res_desc.res.linear.devPtr = (CUdeviceptr)hImageMem;
+        image_res_desc.res.linear.format = format;
+        image_res_desc.res.linear.numChannels = NumChannels;
+        image_res_desc.res.linear.sizeInBytes =
+            pImageDesc->width * PixelTypeSizeBytes * NumChannels;
+      } else if (pImageDesc->type == UR_MEM_TYPE_IMAGE2D) {
+        image_res_desc.resType = CU_RESOURCE_TYPE_PITCH2D;
+        image_res_desc.res.pitch2D.devPtr = (CUdeviceptr)hImageMem;
+        image_res_desc.res.pitch2D.format = format;
+        image_res_desc.res.pitch2D.numChannels = NumChannels;
+        image_res_desc.res.pitch2D.width = pImageDesc->width;
+        image_res_desc.res.pitch2D.height = pImageDesc->height;
+        image_res_desc.res.pitch2D.pitchInBytes = pImageDesc->rowPitch;
+      } else if (pImageDesc->type == UR_MEM_TYPE_IMAGE3D) {
+        // Cannot create 3D image from USM.
+        return UR_RESULT_ERROR_INVALID_VALUE;
+      }
+    } else {
+      // Unknown image memory type.
+      return UR_RESULT_ERROR_INVALID_VALUE;
+    }
+
+    RetErr = urTextureCreate(hContext, hSampler, pImageDesc, image_res_desc,
```

Comment:
Is this function allow to fail, or does it need to be wrapped in: `UR_CHECK_ERROR`?

Reply:
I've wrapped it.